core/src/main/scala/org/apache/spark/CacheManager.scala

```diff
@@ -47,9 +49,13 @@ private[spark] class CacheManager(blockManager: BlockManager) extends Logging {
         val inputMetrics = blockResult.inputMetrics
         val existingMetrics = context.taskMetrics
           .getInputMetricsForReadMethod(inputMetrics.readMethod)
-        existingMetrics.addBytesRead(inputMetrics.bytesRead)
+        existingMetrics.incBytesRead(inputMetrics.bytesRead)
 
-        new InterruptibleIterator(context, blockResult.data.asInstanceOf[Iterator[T]])
+        val iter = blockResult.data.asInstanceOf[Iterator[T]]
+        new InterruptibleIterator(context, AfterNextInterceptingIterator(iter, (next: T) => {
+          existingMetrics.incRecordsRead(1)
```

Comment:
Thinking about it more, we don't really even need the variable to be volatile - the worst that can happen is that the incremental metric we report is slightly out of date.